### ISSUE TITLE
rc.1 external OIDC+LDAP setup, error on new user first login

### ISSUE THREAD

**Issue by NexZhu** [owncloud/ocis#5045]
## Describe the bug

I'm testing `rc.1` with external OIDC+LDAP setup, when a new user try to login, `/ocs/v1.php/cloud/user` returned 500 error.

`proxy` service error log:
```
2022-11-13T13:51:24Z ERR Could not add default role error="{\"id\":\"ocis-settings\",\"code\":403,\"detail\":\"user has no role management permission\",\"status\":\"Forbidden\"}" service=proxy
2022-11-13T13:51:24Z ERR Could not get user by claim error="{\"id\":\"ocis-settings\",\"code\":403,\"detail\":\"user has no role management permission\",\"status\":\"Forbidden\"}" service=proxy
```

Seems like a bug, user shouldn't need  a role management permission to be assigned the `default` role.

(The one user set in the helm chart `adminUUID` option can login as the admin role)

**Comment by NexZhu** [comment]
Another problem which may be related. When I logined as the `adminUUID` user and go to `User management`, there is no group in the beginning, when I tried to create a group, the request ·/graph/v1.0/groups· returned 500 error:
```json
{"error":{"code":"generalException","innererror":{"date":"2022-11-13T14:03:52Z","request-id":"78fac6ae-09f0-4ef4-824f-30aba4a7442b"},"message":"notSupported"}}
```

`graph` service error log:
```
2022-11-13T14:05:12Z DBG No roles in context, contacting settings service service=graph userid=xxx
2022-11-13T14:05:12Z INF calling post group request-id=84de7a53-2eba-4988-88b9-01cba452b8c8 service=graph
2022-11-13T14:05:12Z DBG could not create group: backend error group=null request-id=84de7a53-2eba-4988-88b9-01cba452b8c8 service=graph
2022-11-13T14:05:12Z DBG bytes=160 duration=16.86968 method=POST path=/graph/v1.0/groups proto=HTTP/1.1 request-id=84de7a53-2eba-4988-88b9-01cba452b8c8 service=graph status=500
```

**Comment by ScharfViktor** [comment]
> If I login as the admin users, I can list all users in the oC Web accounts management UI but can't change any user roles.

should be `GRAPH_LDAP_SERVER_WRITE_ENABLED` false? when I tried to create users via webUI I got 500 error with message: "server is configured read-only". https://github.com/owncloud/ocis/blob/master/deployments/examples/ocis_ldap/docker-compose.yml#L75

After starting ocis_ldap with `GRAPH_LDAP_SERVER_WRITE_ENABLED: "true"` I could create users or groups.

**Comment by NexZhu** [comment]
But admin user cannot see users from external LDAP, cannot assign their roles either. Without a default role, they cannot login.

**Comment by NexZhu** [comment]
> > After starting ocis_ldap with `GRAPH_LDAP_SERVER_WRITE_ENABLED: "true"` I could create users or groups.
> 
> If you're using graph to provision users they will immediately have a role assigned. So you must do it via the external ldap (so that the user is just there)

How to assign roles to users in the external LDAP? What's the role attribute read by OCIS? Is there any docs about the OCIS LDAP schema?

**Comment by rhafer** [comment]
> But admin user cannot see users from external LDAP, cannot assign their roles either. Without a default role, they cannot login.

Actually exactly that should be possible. The user with the id matching `adminUUID` should be able to list the existing users from the external LDAP server in the ocis user-management. It should also be able to set the roles of those users in there. (Editing any other user attribute will not work however). If that doesn't work in your setup then there still seems to be something wrong in your setup. Can you share values.yaml (or any other means how you applied your overrides for the  helm chart)?

> What's the role attribute read by OCIS?  Is there any docs about the OCIS LDAP schema?

When using an external LDAP server we don't require a specific schema. The roles and the role assignments are never, regardless for external or internal LDAP) stored in LDAP. We use a dedicated service for that which stores the data in the metadata storage.

> I thought OCIS will insert user data into its own database after checking external LDAP, and I can assign roles to users in OCIS after that.

OCIS does currently not maintain it's own database when an external LDAP server is used. If correctly setup you should be able to assign roles even before the first login (see above). 

> And I got the impression from: https://github.com/owncloud/ocis/issues/4150#issuecomment-1183386298 , that external users will get the default user assigned when loggin in for the first time. Please correct me if I was wrong.

You are right. But there was a bug in  rc1 that prevented that from working. It should be fixed once https://github.com/owncloud/ocis/pull/5065 is merged.

hth

**Comment by NexZhu** [comment]
> Actually exactly that should be possible. The user with the id matching adminUUID should be able to list the existing users from the external LDAP server in the ocis user-management. It should also be able to set the roles of those users in there.
> If correctly setup you should be able to assign roles even before the first login (see above).

Wait, how does OCIS list the existing users from the external LDAP before they login for the first time? Will OCIS search with LDAP filter and get all the matching users from external LDAP?

**Comment by rhafer** [comment]
> Wait, how does OCIS list the existing users from the external LDAP before they login for the first time? Will OCIS search with LDAP filter and get all the matching users from external LDAP?

Exactly. If you don't see them it might be a configuration issue. (That's why I asked for your helm config)

**Comment by individual-it** [comment]
maybe related to https://github.com/owncloud/web/issues/7988 & https://github.com/owncloud/web/issues/7986

**Comment by NexZhu** [comment]
OpenLDAP logs printed when I clicked **User management**:
```log
openldap  | 63775677 conn=1615 op=19226 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(uid=admin))"
openldap  | 63775677 conn=1615 op=19226 SRCH attr=cn uid mail uid uidNumber gidNumber
openldap  | 63775677 conn=1615 op=19226 SEARCH RESULT tag=101 err=0 nentries=1 text=
openldap  | 63775677 conn=1615 op=19227 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(member=uid=admin,ou=people,dc=example,dc=com))"
openldap  | 63775677 conn=1615 op=19227 SRCH attr=cn
openldap  | 63775677 <= mdb_equality_candidates: (member) not indexed
openldap  | 63775677 conn=1615 op=19227 SEARCH RESULT tag=101 err=0 nentries=7 text=
openldap  | 63775677 conn=1615 op=19228 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(|(?=undefined)(?=undefined)(?=undefined)(?uid=)))"
openldap  | 63775677 conn=1615 op=19228 SRCH attr=uid uid mail cn uidNumber gidNumber
openldap  | 63775677 conn=1615 op=19228 SEARCH RESULT tag=101 err=0 nentries=0 text=
openldap  | 63775677 conn=1615 op=19229 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(uid=admin))"
openldap  | 63775677 conn=1615 op=19229 SRCH attr=cn uid mail uid uidNumber gidNumber
openldap  | 63775677 conn=1615 op=19229 SEARCH RESULT tag=101 err=0 nentries=1 text=
openldap  | 63775677 conn=1615 op=19230 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(member=uid=admin,ou=people,dc=example,dc=com))"
openldap  | 63775677 conn=1615 op=19230 SRCH attr=cn
openldap  | 63775677 <= mdb_equality_candidates: (member) not indexed
openldap  | 63775677 conn=1615 op=19230 SEARCH RESULT tag=101 err=0 nentries=7 text=
openldap  | 63775677 conn=1616 op=8 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(|(?=undefined)(?=undefined)(?=undefined)(?cn=)))"
openldap  | 63775677 conn=1616 op=8 SRCH attr=description cn mail cn gidNumber
openldap  | 63775677 conn=1616 op=8 SEARCH RESULT tag=101 err=0 nentries=0 text=
openldap  | 63775677 conn=1615 op=19231 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(uid=admin))"
openldap  | 63775677 conn=1615 op=19231 SRCH attr=cn uid mail uid uidNumber gidNumber
openldap  | 63775677 conn=1615 op=19231 SEARCH RESULT tag=101 err=0 nentries=1 text=
openldap  | 63775677 conn=1615 op=19232 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(member=uid=admin,ou=people,dc=example,dc=com))"
openldap  | 63775677 conn=1615 op=19232 SRCH attr=cn
openldap  | 63775677 <= mdb_equality_candidates: (member) not indexed
openldap  | 63775677 conn=1615 op=19232 SEARCH RESULT tag=101 err=0 nentries=7 text=
openldap  | 63775679 conn=3292 op=11203 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=person)(uid=shujunyi)(|(mail=shujunyi@internetapi.cn)(mail=shujunyi@internetapi.cn@*)))"
openldap  | 63775679 conn=3292 op=11203 SRCH attr=uid
openldap  | 63775679 <= mdb_substring_candidates: (mail) not indexed
openldap  | 63775679 conn=3292 op=11203 SEARCH RESULT tag=101 err=0 nentries=1 text=
openldap  | 63775679 conn=3292 op=11204 BIND anonymous mech=implicit ssf=0
openldap  | 63775679 conn=3292 op=11204 BIND dn="uid=shujunyi,ou=people,dc=example,dc=com" method=128
openldap  | 63775679 conn=3292 op=11204 BIND dn="uid=shujunyi,ou=people,dc=example,dc=com" mech=SIMPLE ssf=0
openldap  | 63775679 conn=3292 op=11204 RESULT tag=97 err=0 text=
openldap  | 63775679 conn=3292 op=11205 BIND anonymous mech=implicit ssf=0
openldap  | 63775679 conn=3292 op=11205 BIND dn="cn=mailserver,ou=dsa,dc=example,dc=com" method=128
openldap  | 63775679 conn=3292 op=11205 BIND dn="cn=mailserver,ou=dsa,dc=example,dc=com" mech=SIMPLE ssf=0
openldap  | 63775679 conn=3292 op=11205 RESULT tag=97 err=0 text=
openldap  | 63775679 conn=3292 op=11206 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=person)(uid=shujunyi)(|(mail=shujunyi)(mail=shujunyi@*)))"
openldap  | 63775679 <= mdb_substring_candidates: (mail) not indexed
openldap  | 63775679 conn=3292 op=11206 SEARCH RESULT tag=101 err=0 nentries=1 text=

···

**Comment by xoxys** [comment]
@rhafer @micbar What about the "If I login as the admin users, I can list all users in the oC Web accounts management UI but can't change any user roles." part? Was that fixed in the linked PR as well?

**Comment by micbar** [comment]
Should be, yes. Needs re testing in the next qa cycle.

**Comment by ScharfViktor** [comment]
re-tested. 
- admin creates new user via ldap (user is without role)
- user can login in ocis - fixed
- admin cannot edit the user role in User management  oC Web. he get 500 with `message: "server is configured read-only"` - still relevant

**Comment by xoxys** [comment]
Thanks. Should I create a new issue or re-open this one?

**Comment by micbar** [comment]
> admin cannot edit the user role in User management oC Web. he get 500 with message: "server is configured read-only" - still relevant

Hmm, this is on a different level. IdK where this error message comes from. Needs investigation. I only see this error in the LDAP code, but assigning a role to a user is not an LDAP operation.

**Comment by micbar** [comment]
@ScharfViktor Which request shows this error message?

**Comment by micbar** [comment]
Found it:

```
curl 'https://ocis.ocis-ldap.latest.owncloud.works/graph/v1.0/users/4c510ada-c86b-4815-8820-42cdf82c3d51' \
  -X 'PATCH' \
  -H 'authority: ocis.ocis-ldap.latest.owncloud.works' \
  -H 'accept: application/json, text/plain, */*' \
  -H 'accept-language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7' \
  -H 'authorization: Bearer eyJhbGciOiJQUzI1NiIsImtpZCI6InByaXZhdGUta2V5IiwidHlwIjoiSldUIn0.eyJhdWQiOiJ3ZWIiLCJleHAiOjE2NjkxMjAxMzIsImp0aSI6InNxMzdVbEhZUm1pTGJiRGdENV9seDJNQjdKSFR4Q1lGIiwiaWF0IjoxNjY5MDMzNzMyLCJpc3MiOiJodHRwczovL29jaXMub2Npcy1sZGFwLmxhdGVzdC5vd25jbG91ZC53b3JrcyIsInN1YiI6InpEUUxQRnlWU3BfWW1Zb1RAOHk1Wk5makZld3NnRGp1OXIxZFBCTDJhQl9QWmF2X0hzd19XV1c4U0w4OFpSbjljcXJBOUtMdmpPXy1EeVdiTTFuRHh6QSIsImxnLnQiOiIxIiwic2NwIjoib3BlbmlkIHByb2ZpbGUgZW1haWwiLCJsZy5pIjp7ImRuIjoiQWRtaW4iLCJpZCI6InVpZD1hZG1pbixvdT11c2VycyxkYz1vd25jbG91ZCxkYz1jb20iLCJ1biI6ImFkbWluIn0sImxnLnAiOiJpZGVudGlmaWVyLWxkYXAifQ.TdrxcoerB_fCcvcglOrezsKKy_zcnB3-ZxT3Qb2Jh2lCzbcZ6YMj37BlOVC5iNpnB5MIuGa9lfzgoiaQ405q-5ge9StwxZ3cyi9PGq9dR804OndaRoS8FGMbAuse02Y6g-lGHMnneRooBZvytxJ1GmCn6iCX-r4k6Q1uyo6rFTIntQ7FupHWzMXM6ena2mDkw_g5iq1_tUT1LFHUT3PE6g6XxRNYTrsp-DLrbG7g2OO9Q4VGwrzPJCPCKEKp5KYP3tcKQbutGEX2TlHL1m0oE9MYndP6krVbGgNQ3D042RvYik3BC-lpv6WlX--XmRdP9kMf0vzbxy6ZRWfpvV5AT-M0xHXLWjfTlj8BvKFKrBci-gV5mhvjTT_fNpdXjemSQWVwDvwQ1hkjh12kB56Ec3Zp3e44q9VKpwdhWgdEF1BQHQMX3AVU4wrIl3oLb-d_amQYkCo9UBdicGvU4HnjK_zQKLzRC_tlWAVGT5l46CUcy0g-VGqUYtzvNxe1Xf10LUbOfR-s73ubUZf45Z98MxRuDlqsNNUFG3nrwbe2u1sOzKffOZEC1KM6guEAOia_tE02rz7vv8RWj1734lHTOBBrsVSXsGSpwbDVjnkBWrBz8mLTnnlr5dVUcN22SyyDjO9Pt9yQmpOtc4PRdRjQn2sEQdg9uru0JB4Cuh7iTs0' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'origin: https://ocis.ocis-ldap.latest.owncloud.works' \
  -H 'referer: https://ocis.ocis-ldap.latest.owncloud.works/user-management/users' \
  -H 'sec-ch-ua: "Google Chrome";v="107", "Chromium";v="107", "Not=A?Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36' \
  -H 'x-request-id: 2b0c3c33-46db-4635-9ead-83990c3a49fa' \
  -H 'x-requested-with: XMLHttpRequest' \
  --data-raw '{"displayName":"Albert Einstein","id":"4c510ada-c86b-4815-8820-42cdf82c3d51","mail":"einstein@example.org","onPremisesSamAccountName":"einstein","memberOf":[],"passwordProfile":{"password":""}}' \
  --compressed
```

This is weird. Why do we use the GraphAPI for Role Assignments? @kulmann @JanAckermann I thought we need to use the settings API for that purpose

**Comment by ScharfViktor** [comment]
> @ScharfViktor Which request shows this error message?
PATCH request.

Role assignment works. Sorry for the confusion. 
I assigned the user role through the api

```
curl --location --request POST 'https://ocis.owncloud.test/api/v0/settings/assignments-add' \
--data-raw '{"account_uuid": "dbc3dadf-5bc9-4b9d-9205-ea0bfa5bca8c", "role_id": "d7beeea8-8ff4-406b-8fb6-ab2dd81e6b11"}'
```

the PATCH request remains to be solved

**Comment by NexZhu** [comment]
> These two queries seems to be problematic and are finding 0 result:
> 
> ```
> openldap  | 63775677 conn=1615 op=19228 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(|(?=undefined)(?=undefined)(?=undefined)(?uid=)))"
> openldap  | 63775677 conn=1615 op=19228 SRCH attr=uid uid mail cn uidNumber gidNumber
> openldap  | 63775677 conn=1615 op=19228 SEARCH RESULT tag=101 err=0 nentries=0 text=
> 
> openldap  | 63775677 conn=1616 op=8 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(|(?=undefined)(?=undefined)(?=undefined)(?cn=)))"
> openldap  | 63775677 conn=1616 op=8 SRCH attr=description cn mail cn gidNumber
> openldap  | 63775677 conn=1616 op=8 SEARCH RESULT tag=101 err=0 nentries=0 text=
> ```

@micbar User management failing to list any external LDAP users is not solved yet.

**Comment by micbar** [comment]
> > These two queries seems to be problematic and are finding 0 result:
> > ```
> > openldap  | 63775677 conn=1615 op=19228 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(|(?=undefined)(?=undefined)(?=undefined)(?uid=)))"
> > openldap  | 63775677 conn=1615 op=19228 SRCH attr=uid uid mail cn uidNumber gidNumber
> > openldap  | 63775677 conn=1615 op=19228 SEARCH RESULT tag=101 err=0 nentries=0 text=
> > 
> > openldap  | 63775677 conn=1616 op=8 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(|(?=undefined)(?=undefined)(?=undefined)(?cn=)))"
> > openldap  | 63775677 conn=1616 op=8 SRCH attr=description cn mail cn gidNumber
> > openldap  | 63775677 conn=1616 op=8 SEARCH RESULT tag=101 err=0 nentries=0 text=
> > ```
> 
> @micbar User management failing to list any external LDAP users is not solved yet.

Works on ocis.ocis-ldap.latest.owncloud.works where this patch is already deployed

**Comment by xoxys** [comment]
Now I'm confused again... https://github.com/owncloud/ocis/pull/5065 is deployed on `ocis.ocis-ldap.latest.owncloud.works` already? Because in this case, admins can **not** change user roles via UI. 

This doesn't match

> Role assignment works. Sorry for the confusion.

If it works through the API, that's fine, but this has worked through the UI in the past and should work again.

**Comment by micbar** [comment]
> If it works through the API, that's fine, but this has worked through the UI in the past and should work again.

That is weird to me too. Needs some frontend changes.

@kulmann Extra ticket in web?

**Comment by kulmann** [comment]
> > If it works through the API, that's fine, but this has worked through the UI in the past and should work again.
> 
> That is weird to me too. Needs some frontend changes.
> 
> @kulmann Extra ticket in web?

Yes please, thank you

**Comment by micbar** [comment]
Web Ticket https://github.com/owncloud/web/issues/8011

**Comment by rhafer** [comment]
> > These two queries seems to be problematic and are finding 0 result:
> > ```
> > openldap  | 63775677 conn=1615 op=19228 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(|(?=undefined)(?=undefined)(?=undefined)(?uid=)))"
> > openldap  | 63775677 conn=1615 op=19228 SRCH attr=uid uid mail cn uidNumber gidNumber
> > openldap  | 63775677 conn=1615 op=19228 SEARCH RESULT tag=101 err=0 nentries=0 text=
> > 
> > openldap  | 63775677 conn=1616 op=8 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(|(?=undefined)(?=undefined)(?=undefined)(?cn=)))"
> > openldap  | 63775677 conn=1616 op=8 SRCH attr=description cn mail cn gidNumber
> > openldap  | 63775677 conn=1616 op=8 SEARCH RESULT tag=101 err=0 nentries=0 text=
> > ```

This is indeed weird. Looks a bit like a mis-configuration, but it might also be an issue with the helm charts. I need to take a deeper look.

**Comment by rhafer** [comment]
> > > These two queries seems to be problematic and are finding 0 result:
> > > ```
> > > openldap  | 63775677 conn=1615 op=19228 SRCH base="ou=people,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=inetOrgPerson)(|(?=undefined)(?=undefined)(?=undefined)(?uid=)))"
> > > openldap  | 63775677 conn=1615 op=19228 SRCH attr=uid uid mail cn uidNumber gidNumber
> > > openldap  | 63775677 conn=1615 op=19228 SEARCH RESULT tag=101 err=0 nentries=0 text=
> > > 
> > > openldap  | 63775677 conn=1616 op=8 SRCH base="ou=groups,dc=example,dc=com" scope=2 deref=0 filter="(&(objectClass=groupOfNames)(|(?=undefined)(?=undefined)(?=undefined)(?cn=)))"
> > > openldap  | 63775677 conn=1616 op=8 SRCH attr=description cn mail cn gidNumber
> > > openldap  | 63775677 conn=1616 op=8 SEARCH RESULT tag=101 err=0 nentries=0 text=
> > > ```
> 
> This is indeed weird. Looks a bit like a mis-configuration, but it might also be an issue with the helm charts. I need to take a deeper look.

@NexZhu I was able to reproduce the problem. It seems to be caused by a bug in the user provider, which constructs a broken LDAP filter under certain circumstances. I'll open a new issue for this with some background and a possible workaround.
